BLD: Install scripts tests only during inplace

gfyoung opened the pull request:
The `scripts` directory is not visible if the installation is not inplace.

Follow-up to #20061.

cc @WillAyd @datapythonista

WillAyd commented:
Would the other option be to just skip these tests if the scripts folder isn't included? I actually wonder if that would be preferable especially in cases where users install a binary and try to run `pd.test()` in the REPL

gfyoung commented:
@WillAyd : I don't see why we should be packaging these tests when we do a release of pandas to the general public.  That's one reason why I would prefer exclusion.

jreback reviewed (commented):
shouldn’t you just put a skip in the tests themselves?

gfyoung commented:
@jreback : See my answer [above](https://github.com/pandas-dev/pandas/pull/22413#issuecomment-414094889).  The only time these tests should be running is if we do an inplace install.  Otherwise, why should we be packaging those tests in the distribution (e.g. the public releases).

jorisvandenbossche commented:
> Would the other option be to just skip these tests if the scripts folder isn't included?

\+1

jorisvandenbossche commented:
@gfyoung I understand your argument, but I think the added complexity (I would expect to find logic about whether the tests are ran/skipped in the tests itself, our `setup.py` is already very complex) is not worth the few kb's for those tests.

gfyoung commented:
Fair enough.  IMHO, these validation tests should not have been located in the `pandas/tests` directory, but oh well. 🤷‍♂️

datapythonista commented:
Not sure what it involves in the CI configuration, but I'm +1 on moving the tests for the scripts to `scripts/tests` too.

gfyoung commented:
@jreback @jorisvandenbossche : Thoughts on moving the tests?  As much as I know both of you don't want the extra complexity in `setup.py` of "skipping" the copying of certain files during distribution, I am also not a fan of seeing tests fail locally just because the `scripts` isn't packaged.

jorisvandenbossche commented:
> I am also not a fan of seeing tests fail locally just because the scripts isn't packaged.

They wouldn't fail, the idea is that they would be skipped.

That said, I am also fine with having the tests live in `scripts/tests/, if that doesn't complicate the travis set-up too much to also run those (I suppose now we everywhere only test `pandas` directory?)

WillAyd commented:
I'm -1 on moving to `scripts/tests`. If we are worried about the added size of the new module then we are making micro-optimizations whose benefit would pale in comparison to the effort making sure that we can account for tests that live outside of the `tests` directory

gfyoung commented:
> whose benefit would pale in comparison to the effort making sure that we can account for tests that live outside

@WillAyd : I'm not sure we can make this assertion yet without giving it a shot first. 😉

gfyoung committed:
TST: Skip scripts test if scripts doesn't exist

If the pandas is not inplace, the scripts directory
will not exist, and the tests will fail.

Follow-up to gh-20061.

jreback commented on pandas/tests/scripts/test_validate_docstrings.py:
```diff
@@ -448,22 +448,37 @@ class TestValidator(object):
+        Import the validation scripts from the scripts directory.
+
+        Because the scripts directory is above the top level pandas package,
+        we need to hack `sys.path` to know where to find it for import.
```
avoid the term 'hack'

jreback commented on pandas/tests/scripts/test_validate_docstrings.py:
```diff
@@ -448,22 +448,37 @@ class TestValidator(object):
+            globals()[global_validate_one] = validate_one
+        except ImportError:
+            # Import will fail if the pandas installation is not inplace.
+            raise pytest.skip("scripts directory does not exist")
```
can you say pandas/scripts

gfyoung replied:
Done.

WillAyd commented on pandas/tests/scripts/test_validate_docstrings.py:
```diff
@@ -448,22 +448,37 @@ class TestValidator(object):
+            globals()[global_validate_one] = validate_one
+        except ImportError:
+            # Import will fail if the pandas installation is not inplace.
+            raise pytest.skip("pandas/scripts directory does not exist")
```
Not sure if this causes an issue but from what I've tested locally if using the imperative skip `validate_docstrings` stays in globals as does the modification to system path. Perhaps we should ensure that code gets executed in the `except` block here?

gfyoung replied:
* Fair point regarding system path modification.
* Not sure how `validate_docstrings` stays in globals if the import fails...

gfyoung replied:
@WillAyd : Addressed the change in follow-up (#22456)

jreback commented:
let's see if this fixes the wheels

jreback commented:
ok wheels fixed: https://travis-ci.org/MacPython/pandas-wheels/builds/418513881
